starlite/datastructures/provide.py

```diff
@@ -68,3 +83,114 @@ def __eq__(self, other: Any) -> bool:
             and other.value == self.value
         )
+
+
+class DependencyCleanupGroup:
+    """Wrapper for generator based dependencies.
+
+    Simplify cleanup by wrapping `next`/`anext` calls in `BackgroundTasks` and providing facilities to `throw` /
+    `athrow` into all generators consecutively. An instance of this class can be used as a contextmanager, which will
+    automatically throw any exceptions into its generators. All exceptions caught in this manner will be re-raised after
+    they have been thrown in the generators.
+    """
+
+    __slots__ = ("_generators", "_closed")
+
+    def __init__(self, generators: Optional[List["AnyGenerator"]] = None) -> None:
+        """Initialize `DependencyCleanupGroup`.
+
+        Args:
+            generators: An optional list of generators to be called at cleanup
+        """
+        self._generators = generators or []
+        self._closed = False
+
+    def add(self, generator: Union[Generator[Any, None, None], AsyncGenerator[Any, None]]) -> None:
+        """Add a new generator to the group.
+
+        Args:
+            generator: The generator to add
+
+        Returns:
+            None
+        """
+        if self._closed:
+            raise RuntimeError("Cannot call cleanup on a closed DependencyCleanupGroup")
+        self._generators.append(generator)
+
+    def _wrap_next(self, generator: "AnyGenerator") -> Callable[[], Coroutine[None, None, None]]:
+        if isasyncgen(generator):
+
+            async def wrapped_async() -> None:
+                await async_next(generator, None)  # type: ignore[arg-type]
+
+            return wrapped_async
+
+        def wrapped() -> None:
+            next(generator, None)  # type: ignore[arg-type]
+
+        return AsyncCallable(wrapped)
+
+    async def cleanup(self) -> None:
+        """Execute cleanup by calling `next` / `anext` on all generators.
+
+        If there are multiple generators to be called, they will be executed in a `TaskGroup`.
+
+        Returns:
+            None
+        """
+        if self._closed:
+            raise RuntimeError("Cannot call cleanup on a closed DependencyCleanupGroup")
+        self._closed = True
+        if len(self._generators) == 1:
+            await self._wrap_next(self._generators[0])()
+        elif self._generators:
+            async with create_task_group() as task_group:
+                for generator in self._generators:
+                    task_group.start_soon(self._wrap_next(generator))
+
+    def wrap_asgi(self, app: "ASGIApp") -> "ASGIApp":
+        """Wrap an ASGI callable such that all generators will be called before it.
+
+        Args:
+            app: An ASGI callable
+
+        Returns:
+            An ASGI callable
+        """
+
+        async def wrapped(scope: "Scope", receive: "Receive", send: "Send") -> None:
+            await self.cleanup()
+            await app(scope, receive, send)
+
+        return wrapped
+
+    async def __aenter__(self) -> None:
```

Comment:
?

Reply:
Well it has to have an `__aenter__` method if it has an `__aexit__` method? The use as a context manager is just easier than manually catching exceptions, throwing them into the generators and re-raising.